TemplePlus/critter.cpp

```diff
@@ -710,6 +710,101 @@ StandPoint LegacyCritterSystem::GetStandPoint(objHndl critter, StandPointType ty
 	return result;
 }
 
+
+int LegacyCritterSystem::GetWaypointsCount(objHndl critter)
+{
+	return (int)objects.getArrayFieldInt64(critter, obj_f_npc_waypoints_idx, 0);
+}
+
+bool LegacyCritterSystem::GetWaypoint(objHndl critter, int index, Waypoint& wp)
+{
+	if (index >= GetWaypointsCount(critter))
+		return false;
+	union conv32
+	{
+		uint32_t u32; // here_write_bits
+		float    f32; // here_read_float
+	} u;
+
+	auto arrSize = objects.getArrayFieldInt64Size(critter, obj_f::obj_f_npc_waypoints_idx);
+	if (arrSize <= 2 + index * 8 + 7)
+		return 0;
+
+	int arr[16] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+	for (int row = 0; row < 8; row++)
+	{
+		int64_t v = objects.getArrayFieldInt64(critter, obj_f::obj_f_npc_waypoints_idx, 2 + index * 8 + row);
+		int32_t y = (int32_t)(v & 0xFFFFFFFFLL);
+		int32_t x = (int32_t)((v & 0xFFFFFFFF00000000LL) >> 32);
+		arr[row * 2] = y;
+		arr[row * 2 + 1] = x;
+	}
+	wp.flags = arr[0];
+	wp.location.location.locx = arr[1];
+	wp.location.location.locy = arr[2];
+	u.u32 = arr[3];
+	wp.location.off_x = u.f32;
+	u.u32 = arr[4];
+	wp.location.off_y = u.f32;
+	u.u32 = arr[5];
+	wp.rotation = u.f32;
+	wp.delay = arr[8];
+
+	wp.anims[0] = (arr[6] & 0x000000ff);
```

Comment:
Why the bitwise operations? It's just a uint8_t array

Reply:
I wanted it to be visibly aligned:
`	wp.anims[0] = (arr[6] & 0x000000ff);
	wp.anims[1] = (arr[6] & 0x0000ff00) >> 8;
	wp.anims[2] = (arr[6] & 0x00ff0000) >> 16;
	wp.anims[3] = (arr[6] & 0xff000000) >> 24;
	wp.anims[4] = (arr[7] & 0x000000ff);
	wp.anims[5] = (arr[7] & 0x0000ff00) >> 8;
	wp.anims[6] = (arr[7] & 0x00ff0000) >> 16;
	wp.anims[7] = (arr[7] & 0xff000000) >> 24;`

Reply:
@DudeMcDude, should it stay, or should I rewrite it somehow?